Pipenets

Aranclanos opened the pull request:
### Purpose
Adds a WIP pipenet system, with two new objects, pipes and connectors.

### Open Questions and Pre-Merge TODOs

- [X]  This fix is tested on the same branch it is PR'ed to.
- [ ]  I correctly commented my code
- [X]  My code is indented with tabs and not spaces
- [X]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [X]  This PR does not bring up any new compile errors
- [X]  This PR has been tested in editor
- [ ]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

### Notes:
Known issues:
-pipes get sucked/pushed by environmental air
-cant rotate pipes
-all pipes are north/south
-advanced pipes (vents, scrubbers, connectors) connect both sides even if their sprite is for south
-pipe sprites go over canisters
-canisters have no sprite to indicate the connection to the connector
-canisters can be pushed or pulled when connected
-pipes can be pulled while anchored
-map lacks pipes, vents and scrubbers don't spawn anchored
-no comments
-clients don't update pipe sprites
-there's nothing to check for a shuttle moving/rotating
~~there are some debug variables~~

Aranclanos committed:
First version of pipenets, work in progress

Aranclanos committed:
Merge branch 'develop' of https://github.com/unitystation/unitystation into pipenets

Aranclanos committed:
Finishes pipenets core behaviour

Aranclanos committed:
Vents, scrubbers and connectors will now be part of the pipenet members.
Adds connectors.
Updates vents and scrubbers with the new pipenet system.
Renames TankInteract.cs to Canister.cs.
Updates pipe sprites to their latest version.
Adds a merge function for GasMix.

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs:
```diff
@@ -0,0 +1,208 @@
+	public Pipenet pipenet;
+	public float volume = 70;
+
+	public int ARANpipenetmembers;
```
lol, why would you add your nickname to field names?

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/AirVent.cs:
```diff
@@ -1,55 +1,51 @@
+			GasMix gasMix = pipenet.gasMix;
+			pipenet.gasMix = gasMix / 2;
+			metaNode.GasMix = metaNode.GasMix + gasMix;
+			metaDataLayer.UpdateSystemsAt(Vector3Int.RoundToInt(transform.position));
```
You're using `transform.position` *everywhere*. **Don't!** Instead, cache a reference to RegisterTile in Pipe class and use `registerTile.WorldPositionServer`, they're already rounded and reliable. transform.position is clientside fluff, never rely on it if you have RegisterTile.

fomalsd replied:
<img width="846" alt="Scr 2019-05-17 в 10 18 12" src="https://user-images.githubusercontent.com/10403536/58136364-78e71880-7c36-11e9-9e1a-fd9ef21cbaac.png">

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs:
```diff
@@ -0,0 +1,208 @@
+	public string ARANpipenetName;
+	public float ARANpipenetVolume;
+
+	public enum Direction
```
you can use powerful `Orientation` struct, by the way

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs:
```diff
@@ -0,0 +1,208 @@
+		foundPipenet.AddPipe(this);
+
+		transform.rotation = new Quaternion();
+		transform.position = Vector3Int.RoundToInt(transform.position);
```
Do they still get blown away by wind? You could take advantage of `IsTileSnap` and `isNotPushable` bools, making sure they can't be blown away or nudged (and snap to tile like lockers and players)

fomalsd reviewed (changes requested):
Not gonna lie, I lol'd about these named fields pretty hard
Check out other comments, there are a few important points

Aranclanos committed:
Removes some leftlover debug variables

Aranclanos committed:
Pipes will now use registerTile to get their location

fomalsd commented on UnityProject/Assets/Scripts/Objects/Canister.cs:
```diff
@@ -28,7 +30,7 @@ public override bool Interact(GameObject originator, Vector3 position, string ha
 			if(connector == null)
 			{
-				var foundConnectors = MatrixManager.GetAt<Connector>(position.RoundToInt(), true);
+				var foundConnectors = MatrixManager.GetAt<Connector>(registerTile.PositionServer, true);
```
MatrixManager works with World positions only – it iterates through all matrices (And `Matrix` class is for local positions), use `WorldPositionServer` here

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs:
```diff
@@ -0,0 +1,209 @@
+		}
+	}
+
+	public List<Vector3Int> GetAdjacentTurfs()
```
this looks redundant, `MetaUtils.GetNeighbors()` probably does what you need

Aranclanos committed:
Fixes canisters using the wrong position to find a connector.
Removes GetAdjacentTurfs() from pipes and replaces it with MetaUtils.GetNeighbors()
